rules/terraform_module_shallow_clone.go

```diff
@@ -0,0 +1,153 @@
+type TerraformModuleShallowCloneRule struct {
+	tflint.DefaultRule
+
+	attributeName string
+}
+
+// NewTerraformModuleShallowCloneRule returns new rule with default attributes
+func NewTerraformModuleShallowCloneRule() *TerraformModuleShallowCloneRule {
+	return &TerraformModuleShallowCloneRule{
+		attributeName: "source",
+	}
+}
+
+// Name returns the rule name
+func (r *TerraformModuleShallowCloneRule) Name() string {
+	return "terraform_module_shallow_clone"
+}
+
+// Enabled returns whether the rule is enabled by default
+func (r *TerraformModuleShallowCloneRule) Enabled() bool {
+	return true
+}
+
+// Severity returns the rule severity
+func (r *TerraformModuleShallowCloneRule) Severity() tflint.Severity {
+	return tflint.WARNING
+}
+
+// Link returns the rule reference link
+func (r *TerraformModuleShallowCloneRule) Link() string {
+	return project.ReferenceLink(r.Name())
+}
+
+// Check checks if Git-hosted Terraform modules use shallow cloning
+func (r *TerraformModuleShallowCloneRule) Check(rr tflint.Runner) error {
+	runner := rr.(*terraform.Runner)
+
+	path, err := runner.GetModulePath()
+	if err != nil {
+		return err
+	}
+	if !path.IsRoot() {
+		// This rule does not evaluate child modules.
+		return nil
+	}
+
+	calls, diags := runner.GetModuleCalls()
+	if diags.HasErrors() {
+		return diags
+	}
+
+	for _, call := range calls {
+		if err := r.checkModule(runner, call); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *TerraformModuleShallowCloneRule) checkModule(runner tflint.Runner, module *terraform.ModuleCall) error {
+	source, err := getter.Detect(module.Source, filepath.Dir(module.DefRange.Filename), []getter.Detector{
+		// https://github.com/hashicorp/terraform/blob/51b0aee36cc2145f45f5b04051a01eb6eb7be8bf/internal/getmodules/getter.go#L30-L52
+		new(getter.GitHubDetector),
+		new(getter.GitDetector),
+		new(getter.BitBucketDetector),
+		new(getter.GCSDetector),
+		new(getter.S3Detector),
+		new(getter.FileDetector),
+	})
+	if err != nil {
+		return err
+	}
+
+	u, err := url.Parse(source)
+	if err != nil {
+		return err
+	}
+
+	// Only check Git-based sources
+	if u.Scheme != "git" {
+		return nil
+	}
+
+	if u.Opaque != "" {
+		// for git:: pseudo-URLs, Opaque is :https, but query will still be parsed
+		query := u.RawQuery
+		u, err = url.Parse(strings.TrimPrefix(u.Opaque, ":"))
+		if err != nil {
+			return err
+		}
+
+		u.RawQuery = query
+	}
+
+	if u.Hostname() == "" {
+		return nil
+	}
+
+	query := u.Query()
+
+	// Check if module is pinned to a specific version
+	ref := query.Get("ref")
+
+	// Skip if not pinned at all
+	if ref == "" {
+		return nil
+	}
+
+	// Skip if it's a raw git commit (40 character hex string)
+	if ref != "" && r.isGitCommit(ref) {
+		return nil
+	}
+
+	// Check if depth parameter is already set
+	if query.Get("depth") != "" {
+		return nil
+	}
+
+	// Only warn for pinned modules that don't have shallow cloning enabled
+	if ref != "" {
+		return runner.EmitIssue(
```

Comment:
You should support auto-fix here and compute the new URL with the parameter added. Then callers that set `--fix` can roll this out without any manual intervention.